### PR TITLE
Implement Error Hooks

### LINKED ISSUE #1522: Proposal: Error hooks
The idea behind error hooks is to execute arbitrary shell commands when a terraform command returns an error. 

Example syntax would be similar to `before_hook` and `after_hook` with required properties `commands` and `execute` and instead of `run_on_error` a different property `on_errors` would be required that would describe a list of errors on which should the hook run, described in the same manner as `retryable_errors`.

An example configuration that imports would run a shell script with `terraform import`:
```hcl
terraform {
  error_hook "import_resource" {
    commands  = ["apply"]
    execute   = ["./import_known_resource.sh"]
    on_errors = [
      "(?s).*A resource with the ID .* already exists.*",
    ]
  }
}
```

The idea behind this proposal is that there are some resources on terraform-providers that create an 'umbrella' resource which can create some child-resources on the cloud provider which are not being represented as a terraform resource. In some cases, the default child-resource configuration is ok, but when that resource needs to be modified, it becomes problematic.

One example of this would be an `azurerm` resource `azurerm_api_management` creates 2 `azurerm_api_management_product` instances which have one `azurerm_api_management_product_policy` each. The `product` can be accessed as `data` but the `product_policy` needs to be imported to state in order to be accessible from terraform.

### PR DESCRIPTION
**Pull request by smitthakkar96**
Fixes #1522

### CODE REVIEW

**Comment by smitthakkar96** [comment]
@yorinasub17 @Ben10k please review

File: cli/cli_app.go
Comment (denis256) [inline cli/cli_app.go:615]: I think `pre-commit run --all-files` should be executed
```
TerraformCommand)<space>{
```

File: config/config.go
Comment (denis256) [inline config/config.go:353]: `camelCase` is preferred over `snake_case`

File: test/fixture-hooks/before-after-and-on-error/terragrunt.hcl
Comment (denis256) [inline test/fixture-hooks/before-after-and-on-error/terragrunt.hcl:30]: Documentation should be updated to reference `error_hook`
https://github.com/gruntwork-io/terragrunt/blob/master/docs/_docs/04_reference/config-blocks-and-attributes.md

**Comment by smitthakkar96** [comment]
@denis256 addressed your comments

**Review by shemul — DISMISSED** [review]

File: docs/_docs/02_features/hooks.md
Comment (denis256) [inline docs/_docs/02_features/hooks.md:68]: I think `error_hook` should be referenced in the "Blocks" list like `after_hook` / `before_hook`, otherwise, users will not know that such block can be used in HCL

https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/

https://github.com/gruntwork-io/terragrunt/blob/master/docs/_docs/04_reference/config-blocks-and-attributes.md

**Review by denis256 — APPROVED** [review]

File: docs/_docs/02_features/hooks.md
Comment (smitthakkar96) [reply docs/_docs/02_features/hooks.md:68]: good point, let me update that

Comment (smitthakkar96) [reply docs/_docs/02_features/hooks.md:68]: updated https://github.com/gruntwork-io/terragrunt/pull/1967/commits/a467f8c6f0ec3f49d9ebf8f6b4800502df71702b

**Review by denis256 — DISMISSED** [review]

**Comment by smitthakkar96** [comment]
@denis256 thanks for the approval, I see you are also the codeowner for this repo. Could please help merging this?

**Comment by smitthakkar96** [comment]
@yorinasub17 @rhoboat @zackproser @brikis98 can we have one more reviewon this as suggested by @denis256 ?

**Comment by rhoboat** [comment]
Reviewing!

File: docs/_docs/02_features/hooks.md
Comment (rhoboat) [inline docs/_docs/02_features/hooks.md:63]: Just some grammatical suggestions here.

```suggestion
Think of error hooks like exception handlers. They allow you to specifiy a list of expressions that can be used to catch errors and run custom commands when those errors occur.
```

File: docs/_docs/02_features/hooks.md
Comment (rhoboat) [inline docs/_docs/02_features/hooks.md:6]: We should update this too.

```suggestion
excerpt: Learn how to execute custom code before or after running Terraform, or when errors occur.
```

File: docs/_docs/04_reference/config-blocks-and-attributes.md
Comment (rhoboat) [inline docs/_docs/04_reference/config-blocks-and-attributes.md:107]:
```suggestion
- `error_hook` (block): Nested blocks used to specify command hooks that run when an error is thrown. The
error must match one of the expressions listed in the `on_errors` attribute.
```

**Comment by smitthakkar96** [comment]
Thanks @rhoboat for your suggestions. Just merged them :)

File: docs/_docs/04_reference/config-blocks-and-attributes.md
Comment (rhoboat) [inline docs/_docs/04_reference/config-blocks-and-attributes.md:202]: Add an extra line here.

```suggestion
  }
  
```

File: docs/_docs/04_reference/config-blocks-and-attributes.md
Comment (rhoboat) [inline docs/_docs/04_reference/config-blocks-and-attributes.md:195]:
```suggestion
  # After an error occurs during apply or plan, run "echo Error Hook executed". This hook is configured so that it will run
  # after any error, with the ".*" expression.
```

File: config/config.go
Comment (rhoboat) [inline config/config.go:361]: Hmm.. so we run the before and after hooks first, then we run the error hooks. We should mention this somewhere in the docs.

**Review by denis256 — DISMISSED** [review]

File: config/config_as_cty_test.go
Comment (rhoboat) [inline config/config_as_cty_test.go:54]: Why is the ErrorHooks block not included under the Terraform one above, so that it would be nested at the same level as the Before/AfterHooks?

File: docs/_docs/02_features/hooks.md
Comment (rhoboat) [inline docs/_docs/02_features/hooks.md:3]: What about this instead?

```suggestion
title: Before, After, and Error Hooks
```

**Review by rhoboat — COMMENTED** [review]
I have some questions.

Does it make sense to always process the error hooks after the other hooks? If you have an after hook that is specified to run on error, what happens? This hook will run first, then the error hook. Is that always the right sequencing? How do we know?

I left a few other nits and questions, but I do think this is a good feature!

File: config/config_as_cty_test.go
Comment (smitthakkar96) [reply config/config_as_cty_test.go:54]: I didn't understand the comment, it is at same level as before/after hooks from what I can see

File: config/config.go
Comment (smitthakkar96) [reply config/config.go:361]: This is addressed https://github.com/gruntwork-io/terragrunt/pull/1967/commits/94037c72168be46d1cb7dbb47d687aea0933148c

**Comment by smitthakkar96** [comment]
@rhoboat good question! Initially I was confused about the execution status I wasn't sure where should I start this dicussion but I think we can deprecate `run_on_error` from before/after hooks in favour of error hooks in future. Why? 
- Because it's very confusing (which one to use)
- Currently if there is no way for the downstream command/script that will be executed by the before/after hooks to know whether it is being executed because of an error or success. If others agree maybe we should add deprecation warning for `run_on_error`. WDYT?

**Comment by rhoboat** [comment]
@smitthakkar96 This turned out to be a long response, so I've made sections. Bear with me! 🐻 

## Deprecating the `run_on_error`
For argument's sake: let's see what's the difference between an after_hook with run_on_error and an error_hook:

```
  after_hook "after_hook_1" {
    commands = ["apply", "plan"]
    execute = ["touch","after.out"]
    run_on_error = true
  }
```

The closest we can get to that with error_hook is

```
  error_hook "error_hook_1" {
    commands  = ["apply", "plan"]
    execute   = ["touch", "after.out"]
    on_errors = [".*"]
  }
```

Are these identical? No, because the after_hook will run regardless of whether there are errors, but the error_hook will only run if there are errors. Taken together, the error_hook is redundant. So I see the error_hook as a more specific way to handle the error-only case.

If you get rid of `run_on_error`, you will have to do:

```
  after_hook "after_hook_1" {
    commands = ["apply", "plan"]
    execute = ["touch","after.out"]
  }
```

plus

```
  error_hook "error_hook_1" {
    commands  = ["apply", "plan"]
    execute   = ["touch", "after.out"]
    on_errors = [".*"]
  }
```
Now, the after_hook makes sure it touches after.out only if there are no errors, and the error_hook makes sure it touches after.out if there are errors.

So I think we can't deprecate the `run_on_error` feature without making it harder to do the same thing. 

## Error/success
> Currently if there is no way for the downstream command/script that will be executed by the before/after hooks to know whether it is being executed because of an error or success.

That is true, however, this the main feature for `run_on_error`. It's for the set of cases where you don't want an error to interrupt execution of a particular hook. Is that still useful / needed? (I would also like to hear from people whether it is.) With the addition of `error_hook` we can handle specific error cases, so I do think this is an important feature.

## Execution order

I've been noodling on this a little more, and I think if we're going with error handling as other languages do it, the ordering should be different. I'll start by describing what I think the intent of each of the hooks is:

`after_hook` : run this thing if there's no errors
`after_hook` with `run_on_error` : always run this thing (like `always` and `finally` in typical exception handling parlance)
`error_hook` : run this thing only if there's some specific errors (or all errors with `".*"`) and it's meant to be used specifically to _handle_ the error. In the docs included for this PR, we're advising that you think of this like error handling.. That is, when you catch an error, you execute the catch block _before_ the finally/always blocks. Right?

Given that, I think the ideal execution order should be the following.

1. `before_hooks`
2. `after_hooks` without `run_on_error`
3. `error_hooks`
4. `after_hooks` with `run_on_error`

I also think this is doable from an implementation standpoint, but will not be as clean as you have it in the PR now.

### More intuitive naming?

The above feels intuitive from an execution standpoint, but it's hard to keep this in mind when writing terragrunt. It would be nice if we could do this, but make it easy to understand. Would naming help? Would it be any better if `run_on_error` for the `after_hook` was renamed to `finally` or `always`? I feel like `run_on_error` is clear, but `always = true` also reads better. `run_on_error` could be taken to mean "only run in case of error", which is wrong. `always` doesn't contain the word "error" but it is very clear that it runs "always" (whether or not there's an error).

But there's still the mental overhead of remembering that an `error_hook` which handles the specific error thrown will run before the `after_hook` in cases of error. Is this really as bad as I'm making it sound? WDYT?

**Comment by smitthakkar96** [comment]
Hey thanks for putting a great deal of thoughts into the execution order. Here is what I think

> That is true, however, this the main feature for `run_on_error`. It's for the set of cases where you don't want an error to interrupt execution of a particular hook. Is that still useful / needed? (I would also like to hear from people whether it is.)

I just realised that I have a such use-case too. During plan regardless of error I want to continue the execution. My use-case was we have a utility that would move state files in S3 incase of renames and path changes (we have folders split per teams `squads/<squad_name>/module`). We rely on `git diff` to detect renames and move operations. When service responsibility changes the modules move from one squad folder to another. What we do with hooks is something like below:

``` hcl
  before_hook "before_plan" {
    commands     = ["plan"]
    execute      = [
      local.migrate_states_binary,
      "--yes",
    ]
    working_dir = local.repo_path
  }
  after_hook "after_plan" {
    commands     = ["plan"]
    execute      = [
      local.migrate_states_binary,
      "--dry-run",
      "--rollback",
    ]
    working_dir = local.repo_path
    run_on_error = true
  }
  before_hook "before_apply" {
    commands     = ["apply"]
    execute      = [
      local.migrate_states_binary,
      "--yes"
    ]
    working_dir = local.repo_path
  }
  error_hook "rollback_on_apply_failure" {
    commands     = ["apply"]
    execute      = [
      local.migrate_states_binary,
     "--rollback",
      "--yes"
    ]
    on_errors = [".*"]
  }
}
```
I guess above the example would look ugly without `run_on_error` where after plan we want to rollback the state changes incase developer doesn't decide to apply.


> The above feels intuitive from an execution standpoint, but it's hard to keep this in mind when writing terragrunt. It would be nice if we could do this, but make it easy to understand. Would naming help? Would it be any better if `run_on_error` for the `after_hook` was renamed to `finally` or `always`? I feel like `run_on_error` is clear, but `always = true` also reads better. `run_on_error` could be taken to mean "only run in case of error", which is wrong. `always` doesn't contain the word "error" but it is very clear that it runs "always" (whether or not there's an error).

I like the idea of renaming it to `always` however it will break backwards compatibility, not sure what is the guideline to do that in `gruntworks`, if changing name is hard maybe we can improve the documentation as a part of this PR explaining when to use `run_on_error`, when to use `error_hooks` and also document the order of execution. 

> Given that, I think the ideal execution order should be the following.

I like the execution order and I think if well documented there won't be mental overhead of remembering that an `error_hook` is executed before `after_hooks` with `run_on_error`. 

Should I work on improving docs and changing the execution order and we can open a separate issue (we can get some more feedback from community, maybe gruntworks can reach out to their customers internally too asking for feedback on naming) to rename `run_on_apply` to `always`?

File: config/config.go
Comment (yorinasub17) [inline config/config.go:361]: NIT: reduce this to the following for DRY:

```go
beforeAndAfterHooks := append(conf.GetBeforeHooks(), conf.GetAfterHooks()...)
allHooks := append(beforeAndAfterHooks, conf.GetErrorHooks()...)
for _, curHook := range allHooks {
```

File: docs/_docs/02_features/hooks.md
Comment (yorinasub17) [inline docs/_docs/02_features/hooks.md:63]: To be consistent:

```suggestion
*Error hooks* are a special type of after hook that act as exception handlers. They allow you to specify a list of expressions that can be used to catch errors and run custom commands when those errors occur. Error hooks are executed after the before/after hooks.
```

**Review by yorinasub17 — COMMENTED** [review]
Thanks for the contribution! This mostly LGTM, but I found a few minor nits, and one bug: the new error hook blocks need to be handled in the respective merge functions for `include`. Specifically:

- https://github.com/gruntwork-io/terragrunt/blob/master/config/include.go#L306
- https://github.com/gruntwork-io/terragrunt/blob/master/config/include.go#L450

**Comment by smitthakkar96** [comment]
@yorinasub17 thanks for your review, apart from nits and fixing the bug, should I also change execution order of hooks like @rhoboat suggested? What are your thoughts on execution order?

**Comment by yorinasub17** [comment]
RE: execution order

I think ideally and what is most intuitive, is if the hooks are executed in the order that it is defined. That is, if you had the following:

```hcl
terraform {
  after_hook "always" {
    commands = ["plan"]
    execute = ["echo", "always"]
    run_on_error = true
  }
  after_hook "on_success" {
    commands = ["plan"]
    execute = ["echo", "success"]
  }
  error_hook "on_error" {
    commands = ["plan"]
    execute = ["echo", "error"]
    on_errors = [".*"]
  }
  after_hook "last" {
    commands = ["plan"]
    execute = ["echo", "last"]
    run_on_error = true
  }
}
```

On error, the output should be:

```
always
error
last
```

and on success, it should be:

```
always
success
last
```

This mimics how terraform `local-exec` provisioners work, and personally would be much easier to reason about because you don't have to know terragrunt internals of how hooks are processed.

This is the disadvantage of introducing a new block, as it makes it hard to know the order of the blocks when they are interleaved.

The only way I think we can implement this is if we use the approach terraform takes of using labels to identify the hook type - E.g.:

```hcl
terraform {
  hook "after" "always" {
    commands = ["plan"]
    execute = ["echo", "always"]
    run_on_error = true
  }
  hook "after" "on_success" {
    commands = ["plan"]
    execute = ["echo", "success"]
  }
  hook "error" "on_error" {
    commands = ["plan"]
    execute = ["echo", "error"]
    on_errors = [".*"]
  }
  hook "after" "last" {
    commands = ["plan"]
    execute = ["echo", "last"]
    run_on_error = true
  }
}
```

That said, this is a rather drastic backward incompatible change, but FWIW, terragrunt is still a pre-1.0 project and we haven't really identified the stable API yet. I think this is at least worth a consideration in an RFC discussion.

In any case, at this stage, I believe the best course of action is to avoid prescribing and mentioning an execution order and leaving that as undefined. That is, the basic execution order of after and error hooks running after the commands is defined and set, but in what order those individual hooks are run is not specified or guaranteed (similar to the unspecified order of iterating through a map in most languages). NOTE: we should document that it is undefined.

Then separately, we should open an RFC discussing what the actual execution order of hooks we want terragrunt to enforce and implement.

That way, we can consider the ideal UX of hook ordering without being tied to implementation details, and also we can be more systematic about the proposal rather than it being implemented/discussed on the fly in a PR.

**Comment by rhoboat** [comment]
Thanks @yorinasub17 for further insights into the execution order! It makes sense to respect the lexical ordering (i.e., the order that the hooks appear in the code). I didn't even think of that. 

I agree with execution order being a separate issue we can tackle later. Let's not block this PR on that. I think it makes good enough sense as it is!

File: config/config_as_cty_test.go
Comment (rhoboat) [reply config/config_as_cty_test.go:54]: @smitthakkar96 Did you make this change? Just wanted to give this thread a bump, since it might have gotten missed!

File: config/config.go
Comment (smitthakkar96) [reply config/config.go:361]: won't work, it will probably fail with error, as `ErrorHook` is a different type. The reason for creating different type of hook as `OnErrors` attribute doesn't need to be in before and after hooks. 

```
cannot use conf.GetErrorHooks() (type []ErrorHook) as type []Hook in append
```

I guess the instead of creating new type `ErrorHook` I will have to use the same type, only disadvantage to that will be OnErrors attribute won't be of any use for other two hooks (i.e before and after we can just add validation I guess).

Comment (yorinasub17) [reply config/config.go:361]: Ah that's right makes sense. You can keep it as is for now then!

Comment (smitthakkar96) [reply config/config.go:361]: I realised `mergeHooks` and `getIndexOfHookWithName` need to be duplicated too, any suggestions for better design/way to do it. Sorry I my relatively new to Golang

File: config/config_as_cty_test.go
Comment (smitthakkar96) [reply config/config_as_cty_test.go:54]: Hey sorry I didn't get the comment, what am I missing here?

File: config/config.go
Comment (yorinasub17) [reply config/config.go:361]: It's out of scope of this PR, but the way to handle this is to define a function `validate` on the `Hook` and `ErrorHook` type, and then define an `interface`. Then, you can construct a list of the interface type, and call the `validate` function in the loop.

File: config/config_as_cty_test.go
Comment (rhoboat) [reply config/config_as_cty_test.go:54]: Ah, I see. I was seeing things indented incorrectly due to the GitHub UI. Looking at the raw file, everything is formatted right. Closing this thread. Sorry for the noise!

**Review by rhoboat — APPROVED** [review]
LGTM!

**Comment by smitthakkar96** [comment]
> bug: the new error hook blocks need to be handled in the respective merge functions for `include`. Specifically:
@yorinasub17 @rhoboat I was about to push a commit to above comment but PR got merged so I created a new PR to handle the merge part https://github.com/gruntwork-io/terragrunt/pull/1982

**Comment by alikhil** [comment]
Hello!

Is it possible to pass catched error to hook now? I checked the code and it seems like it cannot

**Comment by zackproser** [comment]
@alikhil could you please open a new issue and include all the relevant details so that we can better assist you? 

* Which version are you running?
* What command are you typing?
* What output are you expecting? 
* What output are you getting? 

Thank you!

**Pull request merged**
